### PR TITLE
added bottom ad to amp-apester-media

### PR DESCRIPTION
**Pull request by golovatuy**
✨ New feature
added bottom ad block to amp-apester-media campaign

### CODE REVIEW

**Review by OmriKeret — APPROVED** [review]
LGTM

File: extensions/amp-iframely/OWNERS.yaml [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # For an explanation of the OWNERS.yaml rules and syntax, see:
 # https://github.com/ampproject/amp-github-apps/blob/master/owners/OWNERS.example.yaml
 
+- OmriKeret
```
Comment (caroqliu) [inline extensions/amp-iframely/OWNERS.yaml:4]: This addition appears to modify `amp-iframely` which is otherwise not relevant for this PR. Can this change be made in a separate PR please?

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      'height': `${height}`,
+      'type': 'doubleclick',
+      'style':
+        'position: absolute; bottom: 0;left: 50%;margin-left: -150px;margin-bottom: 0;',
+      'layout': 'fixed',
+      'data-slot': `${slot}`,
+      'data-multi-size-validation': 'false',
+    })
+  );
+  ampAd.classList.add('amp-apester-companion');
```
Comment (caroqliu) [inline extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:70]: Should this be a public class that publishers can define styles with? If not please prefix with `i-amphtml-` instead.

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ * @param {Array} bannerSizes
+ * @param {!AmpElement} apesterElement
+ * @return {!Element}
+ */
+function constructCompanionBottomAd(slot, bannerSizes, apesterElement) {
+  const width = bannerSizes[0][0];
+  const height = bannerSizes[0][1];
+  const ampAd = createElementWithAttributes(
+    /** @type {!Document} */ (apesterElement.ownerDocument),
+    'amp-ad',
```
Comment (caroqliu) [inline extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:58]: Do we ensure the `amp-ad` script is installed on the document elsewhere? If not, please do so.

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      'style':
+        'position: absolute; bottom: 0;left: 50%;margin-left: -150px;margin-bottom: 0;',
+      'layout': 'fixed',
+      'data-slot': `${slot}`,
+      'data-multi-size-validation': 'false',
+    })
+  );
+  ampAd.classList.add('amp-apester-companion');
+  apesterElement.appendChild(ampAd);
+  Services.mutatorForDoc(apesterElement).requestChangeSize(ampAd, height);
```
Comment (caroqliu) [inline extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:72]: Do both the append and change size operations cause layout shift?

**Review by caroqliu — CHANGES_REQUESTED** [review]
Had a couple comments/questions. Given this is related to `amp-ad`, would appreciate review from @ampproject/wg-monetization as well :)

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js (hunk 3 above)
Comment (powerivq) [reply extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:58]: Is this intended to be a bottom sticky ad? If so, you should create an element
`<amp-ad sticky=bottom`

**Review by powerivq — CHANGES_REQUESTED** [review]

File: extensions/amp-iframely/OWNERS.yaml (hunk 1 above)
Comment (golovatuy) [reply extensions/amp-iframely/OWNERS.yaml:4]: Yes, we'll place this change in the new PR.

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js (hunk 3 above)
Comment (golovatuy) [reply extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:58]: @powerivq When we use 'sticky=bottom' ad block is sticky to the bottom of the window. It's not good for our case. We want to place ad block to the bottom of the media, but without CSS "position: fixed", which is added by sticky attribute.

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js (hunk 2 above)
Comment (golovatuy) [reply extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js]: Ok, added prefix.

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js (hunk 4 above)
Comment (golovatuy) [reply extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:72]: It doesn't cause any Cumulative Layout Shift (CLS) issue on the page. Media is loaded without shift with the height which is mention in the height attribute of the amp-apester-media element. If original height of media is bigger we show "Full size" button on the center of the media. After clicking on it user will see full size media. When publisher place <amp-apester-media> with bottom ad block it doesn't cause layout shift.

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js (hunk 3 above)
Comment (golovatuy) [reply extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:58]: @caroqliu It's not clear why we need to check amp-ad script is installed. It's a "managed resource" (whether and when it will be loaded and displayed to the user is decided by the AMP runtime).

**Review by powerivq — CHANGES_REQUESTED** [review]

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
+/**
+* Copyright 2019 The AMP HTML Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (golovatuy) [reply extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js]: Fixed, thanks.

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    /**@type {!JsonObject}*/ (media),
+    'campaignData.bottomAdOptions'
+  );
+  if (!bottomAdOptions) {
+    return;
+  }
+  const enabledBottomAd = getValueForExpr(
+    /**@type {!JsonObject}*/ (bottomAdOptions),
+    'enabled'
+  );
```
Comment (caroqliu) [inline extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:36]:
```suggestion
  const enabledBottomAd = getValueForExpr(
    /**@type {!JsonObject}*/ (media),
    'campaignData.bottomAdOptions.enabled'
  );
```

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  if (
+    enabledBottomAd &&
+    bottomAdOptions['videoPlayer'] === ALLOWED_AD_PROVIDER
+  ) {
+    const slot = bottomAdOptions['tag'];
+    const bannerSizes = [[300, 50]];
```
Comment (caroqliu) [inline extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:42]: Is there a reason behind these preset values being chosen? I see `50` corresponds to the hardcoded bottom ad height -- why `300` for width?

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js (hunk 3 above)
Comment (caroqliu) [reply extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:58]: Here is an example:
https://github.com/ampproject/amphtml/blob/9bb603c5803ec809c458ec54ca4cd20f12463a76/extensions/amp-auto-ads/0.1/amp-auto-ads.js#L52-L56

**Review by caroqliu — CHANGES_REQUESTED** [review]
Could you please add examples in `examples/apester-media.amp.html` of the new feature so we can demo the PR changes, as well as unit tests? Thank you!

File: extensions/amp-apester-media/0.1/amp-apester-media.js [hunk 8]
```diff
@@ -317,11 +321,23 @@ class AmpApesterMedia extends AMP.BaseElement {
...
+                  const campaignData = media['campaignData'];
+                  if (campaignData) {
+                    const bottomAdOptions = campaignData['bottomAdOptions'];
+                    if (bottomAdOptions && bottomAdOptions.enabled) {
+                      this.hasBottomAd_ = true;
+                      this.iframe_.contentWindow./*OK*/ postMessage(
+                        /** @type {JsonObject} */ ({
+                          type: HAS_BOTTOM_AD_EVENT,
+                          adHeight: BOTTOM_AD_HEIGHT,
+                        }),
```
[the remark this answers is not in the document]
Comment (golovatuy) [reply extensions/amp-apester-media/0.1/amp-apester-media.js]: @caroqliu, I added unit test for the bottom ad case. About examples, I can not do that, - bottom ad will be shown only for publishers who setting up ad campaign with bottom ad settings. We'd like to show "amp-apester-media" elements without ad blocks in examples.

File: extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js (hunk 7 above)
Comment (golovatuy) [reply extensions/amp-apester-media/0.1/monetization/companion/bottomAd.js:42]: At this moment we use only 300x50 size for bottom ad blocks as a most popular for that placement.

**Review by OmriKeret — APPROVED** [review]

File: extensions/amp-apester-media/0.1/amp-apester-media.js [hunk 9]
```diff
@@ -154,6 +159,11 @@ class AmpApesterMedia extends AMP.BaseElement {
       renderer: true,
       tags: extractTags(this.getAmpDoc(), this.element),
     };
+    const ampdoc = this.getAmpDoc();
+    Services.extensionsFor(this.win)./*OK*/ installExtensionForDoc(
+      ampdoc,
+      AD_TAG
+    );
```
Comment (caroqliu) [inline extensions/amp-apester-media/0.1/amp-apester-media.js:166]: Let's not install `amp-ad` until we know there is a bottom ad enabled (around line 334)

**Review by caroqliu — APPROVED** [review]

File: extensions/amp-apester-media/0.1/amp-apester-media.js (hunk 9 above)
Comment (powerivq) [reply extensions/amp-apester-media/0.1/amp-apester-media.js:166]: +1

Comment (golovatuy) [reply extensions/amp-apester-media/0.1/amp-apester-media.js]: Agree, fixed

**Comment by golovatuy** [comment]
Hi. Thanks for the comments, @powerivq. Fixed last comment.

**Comment by powerivq** [comment]
The tests are blocking it from being merged. Let's do a rebase.

**Comment by golovatuy** [comment]
@powerivq, thanks. Synced with main branch. Ready.

**Comment by golovatuy** [comment]
@powerivq, synced one more time.

**Comment by OmriKeret** [comment]
Hi @powerivq, is there any other change we need to do in order to get this PR merged?

**Comment by OmriKeret** [comment]
Hi @caroqliu, could you help us push this PR forward? 
Maybe someone else can review it?

Thanks!

**Review by powerivq — APPROVED** [review]

**Comment by rsimha** [comment]
Closing and reopening this PR to retrigger CI checks.

**Comment by golovatuy** [comment]
@kristoferbaxter ready to merge, thanks

**Comment by rsimha** [comment]
Merged. Thanks for contributing to AMP!

**Comment by golovatuy** [comment]
> Merged. Thanks for contributing to AMP!

Thank you.

**Comment by OmriKeret** [comment]
Hi @rsimha! 
How can we track the deployment status of this PR?

Thanks!

**Comment by powerivq** [comment]
@OmriKeret It should be live.
